### PR TITLE
Explicit template instantiate variable

### PR DESCRIPTION
**Pull request by OwenArnold**
Move macros to `variable.tcc`

### CODE REVIEW

File: core/include/scipp/core/variable.tcc [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+                              Vector<underlying_type_t<__VA_ARGS__>>);         \
+  template Vector<underlying_type_t<__VA_ARGS__>>                              \
+      &Variable::cast<__VA_ARGS__>(const bool);                                \
+  template const Vector<underlying_type_t<__VA_ARGS__>>                        \
+      &Variable::cast<__VA_ARGS__>(const bool) const;
+
+/**
+  Support explicit instantionatons for templates for generic VariableView
+*/
+#define INSTANTIATE_SLICEVIEW(...)                                             \
```
Comment (SimonHeybrock) [inline core/include/scipp/core/variable.tcc]: I think this should be merged into the other macro.

File: core/test/variable_test.cpp [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #include "scipp/core/dimensions.h"
 #include "scipp/core/except.h"
 #include "scipp/core/variable.h"
+#include "scipp/core/variable.tcc"
```
Comment (SimonHeybrock) [inline core/test/variable_test.cpp]: I would avoid having the `tcc` included in the main test file. After all we want to ensure that the built-in code for `Variable` compiles and runs *without* having that file included.

Move the instantiation tests to a different source file?

File: core/test/variable_test.cpp [hunk 3]
```diff
@@ -90,6 +91,23 @@ TEST(Variable, span_references_Variable) {
   EXPECT_EQ(observer[0], 1.0);
 }
 
+using OnDemandType = int16_t; // Not instanitated in cpp.
```
Comment (SimonHeybrock) [inline core/test/variable_test.cpp]: I'd use a custom struct defined locally. There is a change that `int16_t` is added to the built-ins later, and then we would need to go back and fix this test.

File: core/test/variable_test.cpp [hunk 4]
```diff
@@ -467,8 +485,8 @@ TEST(Variable, broadcast) {
   EXPECT_EQ(broadcast(var, {Dim::Y, 2}), var);
   EXPECT_EQ(broadcast(var, {{Dim::Y, 2}, {Dim::X, 2}}), var);
 
-  // No transpose done, should this fail? Failing is not really necessary since
-  // we have labeled dimensions.
+  // No transpose done, should this fail? Failing is not really necessary
+  // since we have labeled dimensions.
```
Comment (SimonHeybrock) [inline core/test/variable_test.cpp]: Accidental code format changes here and below?

File: core/include/scipp/core/variable.tcc [hunk 5]
```diff
@@ -751,9 +751,10 @@ VariableView<underlying_type_t<T>> VariableProxy::castVariances() const {
   return requireT<DataModel<Vector<TT>>>(data()).variancesView(dims());
 }
 /**
-  Support explicit instantionatons for templates for generic Variable
+  Support explicit instantiations for templates for generic Variable and
+  VariableConstProxy
 */
-#define INSTANTIATE_VARIABLE(...)                                              \
+#define INSTANTIATE_VARIABLE_AND_SLICE(...)                                    \
```
Comment (SimonHeybrock) [inline core/include/scipp/core/variable.tcc]: I'd keep the name `INSTANTIATE_VARIABLE`. This is in principle user-facing and should not be loaded with technical details.

Comment (OwenArnold) [reply core/include/scipp/core/variable.tcc]: That's what I had originally for the same reasons you said. I'll change it back.

**Pull request merged**